Check for the integrity of the data directory

Learath2 opened the pull request:
Isn't finished yet but it'd be nice if you could take a look before I spend more time on it.

Fixes #2748

def- commented on scripts/data_hash.py:
```diff
@@ -0,0 +1,61 @@
+from hashlib import sha256
+
+files = []
+for p in Path("data/").rglob('*'):
```
Probably also autoexec_server.cfg

Learath2 replied:
Nope, sorry. Must've slipped through

def- commented on CMakeLists.txt:
```diff
@@ -1379,6 +1380,13 @@ generate_source("src/game/generated/protocol.h" "network_header")
 generate_source("src/game/generated/server_data.cpp" "server_content_source")
 generate_source("src/game/generated/server_data.h" "server_content_header")
 
+add_custom_command(OUTPUT "src/engine/generated/data_hash.h"
```
(Why not a cpp file? Probably doesn't matter much if it's only included in one cpp, but might become more annoying if more include it for some reason.)

def- commented on src/game/client/components/menus.cpp:
```diff
@@ -1219,6 +1221,14 @@ int CMenus::Render()
 		}
+		else if(m_Popup == POPUP_DATA_INTEGRITY)
+		{
+			pTitle = Localize("Data directory dirty");
```
Data Directory Dirty?

def- commented on src/game/client/components/menus.cpp:
```diff
@@ -1219,6 +1221,14 @@ int CMenus::Render()
+		{
+			pTitle = Localize("Data directory dirty");
+			//TODO: Get and print the modified files, it'd be wasteful to get this tick so cache it
+			pExtraText = Localize("There are files in the data directory that have been modified.\nThese files will be removed on updates.\nWe suggest you move them into config_directory.");
```
Maybe without explicit \n and have it wrap on its own. Please str_append the list of files too.

Wording: "You have changed files, they may be overwritten by future updates. Customize the client by putting your own files into the config directory instead:"

def- commented on src/game/client/components/menus.cpp:
```diff
@@ -1219,6 +1221,14 @@ int CMenus::Render()
+			pTitle = Localize("Data directory dirty");
+			//TODO: Get and print the modified files, it'd be wasteful to get this tick so cache it
+			pExtraText = Localize("There are files in the data directory that have been modified.\nThese files will be removed on updates.\nWe suggest you move them into config_directory.");
+			pButtonText = Localize("Ok");
```
Maybe have a button to add the current files to some ignore list? But might be too annoying to maintain and too much work. I guess a compromise would be a config variable `cl_check_data_integrity 0`.

We have the full files on update6.ddnet.tw. Can you offer a button `Reset Files` to fix them? If you want to go even fancier you could even move the customized files to config directory automatically on button click and then download the files from update6 to replace them in our data dir.

def- reviewed (changes requested):
I really like it. Thanks for working on this! We should have come up with this years ago. I somehow only realized it after I saw the similar feature in Steam. I guess it's hard to come up with original ideas even when the problem stares at you for years.

Ekidoxx commented:
Would it be possible to automatically move the custom files to config_directory, or optionally with a button added to the warning popup?

def- commented:
I suggested something similar:

> We have the full files on update6.ddnet.tw. Can you offer a button Reset Files to fix them? If you want to go even fancier you could even move the customized files to config directory automatically on button click and then download the files from update6 to replace them in our data dir.

def- commented:
@Learath2 Are you still interested in working on this?

Learath2 commented:
I will finish this up. I just had a sudden burst of irl things to handle. I'll have a little more time starting friday. 

If you want it before then, you could take over.

def- commented:
Nope, that's totally fine!

heinrich5991 commented:
First impressions:

- Probably shouldn't load all the data in the main thread, IO is slow and this will increase startup time.
- Maybe move the generated data to generated instead of engine/generated, game/generated, we use the game/generated from engine already and Teeworlds also moved it there.

Learath2 commented:
We could move this to a thread, but then there is a bit of an issue with when do we warn the player. I really want the warning to come before they ever join a server. So I could wait for the integrity check to be done before letting the client connect, but is that really worth it? They would only be able to browse servers/settings while waiting for it.

The extra startup time we can make feel more responsive by moving the loading bar

I will move the data to game/generated. My initial idea was going to lead to more generated files that would never be used from the game so I just set it up this way.

def- commented:
Move it to a thread, start it at the start of initialization, at end of initialization wait for it to be done? Then at least it's in parallel.

heinrich5991 commented:
We can show a warning via the warning system Jupeyy introduced, even much later after starting.

I don't think a longer loading time is acceptable for this feature which most users don't need.

Learath2 commented:
We can show the warning whenever we want, question is is that good UX?

heinrich5991 commented:
IMO the UX of the common case (the user has not meddled with their data directory) should take priority.

Learath2 commented:
TODO:
- [ ] Popup text
- [ ] `CStorage::Store()`
- [ ] More error checking?

def- commented on CMakeLists.txt:
```diff
@@ -1561,7 +1568,8 @@ set_src(GAME_SHARED GLOB src/game
 )
 # A bit hacky, but these are needed to register all the UUIDs, even for stuff
 # that doesn't link game.
```
Is that comment with UUIDs still up-to-date?

def- reviewed (commented):
There is no scroll bar in case many files are affected.
![screenshot-20201010@003231](https://user-images.githubusercontent.com/2335377/95636717-1a6f4400-0a90-11eb-8b9e-502a9cff31fb.png)
I still think it would be great to automatically move the modified files to config_directory to prevent players from losing their modifications. I guess we'll need a way to find config_directory then. But if we find the settings_ddnet.cfg from there, should be doable?

I also think the "Fix" in popup should be more descriptive.

About ignoring: I guess we don't want a "Keep ignoring" checkbox. Only cl_check_data_integrity

def- commented:
If some files fail to download the entire fix is aborted? I think we should still fix what we can.

Learath2 commented:
> I still think it would be great to automatically move the modified files to config_directory to prevent players from losing their modifications. I guess we'll need a way to find config_directory then. But if we find the settings_ddnet.cfg from there, should be doable?
We find `settings_ddnet.cfg` from any available storage dir, not just the config_directory but now that I looked a bit into it, we could use `IStorage::TYPE_SAVE` that's the directory we use anyway.

> I also think the "Fix" in popup should be more descriptive.
"Restore Files" as you first proposed?

> About ignoring: I guess we don't want a "Keep ignoring" checkbox. Only cl_check_data_integrity
We could add one, but I thought it'd encourage people not to deal with the problem.

> If some files fail to download the entire fix is aborted? I think we should still fix what we can.
Definitely, I'll keep a log of the downloads that failed and display that before restarting.

def- commented:
> "Restore Files" as you first proposed?

Sounds good.

Learath2 commented:
Okay I was extremely slow with this, sorry. I added the scrollbars, I added an auto move to config_directory. The UI text is all that remains in my eyes but lmk if you catch anything else.

Learath2 commented:
I really don't know how to get clang-tidy happy

def- commented:
I will try to fix clang-tidy.

Issues found:
- Scroll wheel doesn't work inside of the boxes.
- "Some text describing wtf this popup is about here"
- I get everything shown as extra+missing at the same time.
![screenshot-20210412@132517](https://user-images.githubusercontent.com/2335377/114387292-88f1a380-9b92-11eb-90bb-a3714a1a19d4.png)

def- commented:
> Okay I was extremely slow with this, sorry.

No need to be sorry. You can work on DDNet topics at any speed you want. I'm happy to have developers working on it at all :)

Learath2 commented:
> I will try to fix clang-tidy.
> 
> 
> 
> Issues found:
> 
> - Scroll wheel doesn't work inside of the boxes.
> 
> - "Some text describing wtf this popup is about here"
> 
> - I get everything shown as extra+missing at the same time.
> 
> ![screenshot-20210412@132517](https://user-images.githubusercontent.com/2335377/114387292-88f1a380-9b92-11eb-90bb-a3714a1a19d4.png)
> 
> 

Can you please show me the matching parts of missing and extra? I thought I fixed this by using paths relative to datadir but apparently not.

def- commented:
![screenshot-20210412@142554](https://user-images.githubusercontent.com/2335377/114394009-00c3cc00-9b9b-11eb-9d3c-52cee97019e3.png)

def- commented:
There is no way to make the "ignore" stick to the selected files. Or at least pressing Escape to skip the Data Integrity dialog.

Learath2 commented:
![sshot](https://user-images.githubusercontent.com/490500/115124223-71b31b80-9fc1-11eb-863b-cec17cfc1d80.png)
New screenshot since I altered the looks.

@def- the bug above was a rebase mistake that slipped my check. Should work fine now

def- commented on src/game/client/components/menus.cpp:
```diff
@@ -2619,6 +2646,197 @@ int CMenus::Render()
+			ModifiedHeader.VSplitMid(0, &ModifiedDiscard);
+
+			static int s_ExtraDiscard = 0;
+			if(DoButton_CheckBox(&s_ExtraDiscard, "Discard", s_ExtraDiscard, &ExtraDiscard) && Updater()->State() == IUpdater::CLEAN)
```
Localize("Discard")
But I think discard is not so clear, maybe something like Permanently Ignore?

Learath2 replied:
Discard just deletes the modifications, it's not permanently ignore. I honestly don't know where I would even fit an ignore option...

def- replied:
Ok, fair enough.

def- replied:
Still needs the Localize("Discard") though.

def- commented on src/game/client/components/menus.cpp:
```diff
@@ -2619,6 +2646,197 @@ int CMenus::Render()
+			Box.VMargin(20.0f, &Box);
+			Box.VMargin(20.f / UI()->Scale(), &Text);
+			Text.h = 24.f / UI()->Scale();
+			int LCount = UI()->DoLabelScaled(&Text, Localize("Your data directory is modified. This configuration is not supported, \"Restore Files\" will move modified and extra files to config_directory and fetch missing files."), FontSize / 1.2f, -1, (int)Text.w);
```
What about Extra files?

Learath2 replied:
Huh, the description clearly says modified and extra go to config_directory

def- commented:
Files are downloaded from the updater directory. If we update files but someone doesn't update their client, they would download the newly updated files and client would still show wrong hash, correct? I think we need to store the files for each released version on a ddnet.tw server or alternatively we only enable this integrity check if your version is the most recent version.

heinrich5991 commented:
> Files are downloaded from the updater directory. If we update files but someone doesn't update their client, they would download the newly updated files and client would still show wrong hash, correct? I think we need to store the files for each released version on a ddnet.tw server or alternatively we only enable this integrity check if your version is the most recent version.

A simple solution to this would be to download files by sha256 value instead of filename from the server.

Learath2 commented:
I already restricted this to the latest version. If we want to let older versions use this aswell I think I'd rather re-do the whole updater with it too, which would add another month or two.

def- commented:
> I already restricted this to the latest version.

Good enough for now, thanks.

def- commented:
Escape key and enter should work. Maybe change the message to just show "Do you want to restore XXX and remove YYY?" and show "Yes" and "No" as options.

def- commented:
> > I already restricted this to the latest version.
> 
> Good enough for now, thanks.

Ah, one problem with this is that we can't beta-test the feature, right?

Learath2 commented:
> > > I already restricted this to the latest version.
> > 
> > Good enough for now, thanks.
> 
> Ah, one problem with this is that we can't beta-test the feature in a release candidate, right?

Hm, that is actually quite troubling for a feature that has potential to break a lot of things. Maybe I should take the time to re-do the update system to be capable of arbitrary updates between versions?

def- commented:
Easiest thing to beta test would be to make an updaterc.ddnet.tw and use that in the updater during RC testing.

Learath2 commented:
I think this is as good as this gets. I've also concluded that our storage system needs an overhaul. Using indices for special paths is very fragile.

IDK how best to test this, I've tested a lot of configurations and situations and it shouldn't really break anything, but it very well could.

def- commented:
If your data  directory has been deleted entirely, it will scan some other directory instead, in my case /usr/share/ddnet/data/:
![screenshot-20210528@110312](https://user-images.githubusercontent.com/2335377/119959510-58c85d80-bfa4-11eb-8c50-a98b5e34dd61.png)

But not sure if this can be fixed easily.

def- commented:
Should we get this into 15.5? I could make RC2 with this change. Seems large and I don't want to wait another few months for it.

Learath2 commented:
> If your data directory has been deleted entirely, it will scan some other directory instead, in my case /usr/share/ddnet/data/
> But not sure if this can be fixed easily.

This was my concern with the storage system in general, the way it's set up doesn't let me easily know where I'm at. As much as it pains me maybe we should just scrap this until we have a better storage system :/

def- commented:
One more problem: Can we set development builds not to do the integrity checks?

heinrich5991 commented:
What's the status of this? This is basically finished, some minor problem found by deen above, but other than that it could go in?

Learath2 commented:
Development builds already don't do integrity checks so that's fixed. My main issue was the data directory thing, the game really shouldn't be going around searching for any data directory

def- commented:
Yeah, but it's probably not too bad with looking for data dir. I'd rather have this in now and get improvements later.

heinrich5991 commented:
I guess it's good and we should try to get it in.

Can you rebase it onto latest master?

Learath2 commented:
I'm still scared of this going ham on someones data directory located in some weird location, but if you think it's good enough I'll rebase it

def- commented:
I agree with heinrich, let's get it in

heinrich5991 commented:
Status: Waiting for rebase.

VoxelDoesCode commented:
Is this flexible with newer updates now, since it has been a while? Or would this need to be changed to compensate with what has happened over the past half year?

heinrich5991 commented:
> Is this flexible with newer updates now, since it has been a while? Or would this need to be changed to compensate with what has happened over the past half year?

Should be flexible/no problem.

Status: Still waiting for rebase.

heinrich5991 commented:
Status: Needs rebase.

edg-l commented:
@Learath2 believes this should belong to a launcher
